biolink2/example.rkt

```diff
@@ -54,32 +54,51 @@
                  (mat 'close)
                  (printf "Finished processing ~s rows\n" count)))))
 
-;; materialize the (concept curie property value) relation if not present
-(unless (directory-exists? (db-path "concept"))
-  (printf "buildling relation: ~s; ~s\n"
-          (db-path fnin.nodeprop) (db-path "concept"))
-  (let/files ((in (db-path fnin.nodeprop))) ()
-    (materialize-dsv-stream
-      in header.nodeprop
-      ;; input columns
-      '(curie property value) buffer-size (db-path "concept")
-      ;; attribute names, types, and internal key name for indexing
-      '(curie property value) '(string string string) 'id
-      ;; primary table columns (without any sorted-columns)
-      '(((curie property value) . ())
-        ;; index table columns (without any sorted-columns)
-        ((value property id)    . ())))))
+;; materialize a relation if not present
+(define (materialize-relation name fnin header fields types)
+  (unless (directory-exists? (db-path name))
+    (printf "buildling relation: ~s; ~s\n"
+            (db-path fnin) (db-path name))
+    (let/files ((in (db-path fnin))) ()
+      (materialize-dsv-stream
+        in header
+        ;; input columns
+        fields buffer-size (db-path name)
+        ;; attribute names, types, and internal key name for indexing
+        fields types 'id
+        ;; primary table columns (without any sorted-columns)
+        `((,fields . ())
```

Comment:
This is probably fine for some basic exploration.  In general, we'll want more precise control over layout of the primary table, which will sometimes differ from the intuitive order you'd give to the relation attributes.

Since there are so many arguments, I'm thinking of redefining materializer to take its arguments in a more keyword style, using an association list (not Racket-style keyword args). This would let us incorporate your reasonable default choices while allowing us to override them.